### PR TITLE
Fix Issue 16082 - Can't access alias this member with same name as module

### PR DESCRIPTION
**Pull request by RazvanN7**

### CODE REVIEW

File: src/dmd/expressionsem.d [hunk 1]
```diff
@@ -2200,6 +2200,37 @@ private Module loadStdMath()
     return impStdMath.mod;
 }
 
+/*
```
Comment (stefan-koch-sociomantic) [inline src/dmd/expressionsem.d]: Add a `*` for ddoc
ddoc starts woith `/**`

File: src/dmd/expressionsem.d [hunk 2]
```diff
@@ -2351,22 +2382,10 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
             return;
         }
 
-        if (hasThis(sc))
+        if (auto e = checkAliasThis(exp.loc, sc, exp.ident))
```
Comment (stefan-koch-sociomantic) [inline src/dmd/expressionsem.d]: avoid `if auto =` the inliner can't handle it

**Comment by RazvanN7** [comment]
@UplinkCoder Done.

File: src/dmd/expressionsem.d [hunk 3]
```diff
@@ -2351,22 +2382,10 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
             return;
         }
 
-        if (hasThis(sc))
+        if (Expression e = checkAliasThis(exp.loc, sc, exp.ident))
```
Comment (stefan-koch-sociomantic) [inline src/dmd/expressionsem.d]: I meant putting an `Expression e` outside of the if, assigning it, and then check for `if (e)`

File: src/dmd/statementsem.d [hunk 4]
```diff
@@ -3084,6 +3084,16 @@ else
             if (rs.exp.op == TOK.type)
                 rs.exp = resolveAliasThis(sc, rs.exp);
 
+            if (rs.exp.op == TOK.scope_)
```
Comment (WalterBright) [inline src/dmd/statementsem.d]: This being checked for only in return statements seems awfully specific. What if `modulename` simply appears in an expression? Or other statement type?

**Comment by WalterBright** [comment]
The fix appears to be in the wrong place. It should be part of the normal name lookup - first the members are looked up, then alias this, then names in the scope enclosing the struct.

Because of this bug, I suspect that the alias this lookups are all wrong.

**Comment by RazvanN7** [comment]
@WalterBright 

>The fix appears to be in the wrong place. It should be part of the normal name lookup - first the members are looked up, then alias this, then names in the scope enclosing the struct.

This wasn't at all obvious from the implementation. In ScopeDsymbol.search there is no mention of alias this; when semantic is performed on IdentifierExp first a normal scope search is done (from the current scope to the upper ones, including the imported scopes); once that fails, alias this is tried. From this implementation, I deduced the the order is : local scope, upper scopes, alias this. Are saying this is wrong ?

**Pull request closed without merging**